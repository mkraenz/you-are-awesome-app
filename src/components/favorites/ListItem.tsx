import React, { FC } from "react";
import { useTranslation } from "react-i18next";
import { StyleSheet, View } from "react-native";
import { List, Paragraph, Subheading } from "react-native-paper";
import { IMessageContent } from "../../state/state/IMessage";

interface Props extends IMessageContent {
    isodate: string;
}

const styles = StyleSheet.create({
    description: {
        flex: 1,
        flexDirection: "row",
        justifyContent: "space-between",
    },
    date: {
        fontSize: 10,
    },
});

const ListItem: FC<Props> = ({ author, text, country, isodate }) => {
    const { t } = useTranslation();

    return (
        <List.Item
            accessibilityStates={{}}
            title={<Subheading>{text}</Subheading>}
            titleNumberOfLines={5}
            description={() => (
                <View style={styles.description}>
                    <Paragraph>
                        {author}
                        {t("from")}
                        {country}
                    </Paragraph>
                    <Paragraph style={styles.date}>{isodate}</Paragraph>
                </View>
            )}
        ></List.Item>
    );
};
export default ListItem;
