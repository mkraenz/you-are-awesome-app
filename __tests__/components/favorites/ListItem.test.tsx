import { render } from "@testing-library/react-native";
import { noop } from "lodash";
import React from "react";
import "react-native";
import renderer, { act } from "react-test-renderer";
import ListItem from "../../../src/components/favorites/ListItem";
import TestLocalizationProvider from "../../helpers/TestLocalizationProvider";

it("renders correctly in normal mode", async () => {
    const tree = renderer
        .create(
            <TestLocalizationProvider>
                <ListItem
                    author="Merlin"
                    country="The old Kingdom of Britania"
                    isodate="975-09-17"
                    onLongPress={noop}
                    onPressInSelectMode={noop}
                    selectMode={false}
                    selected={false}
                    text="The story of King Arthur"
                />
            </TestLocalizationProvider>
        )
        .toJSON();
    await act(async () => {});

    const treeStr = JSON.stringify(tree);
    expect(treeStr).toContain("Merlin");
    expect(treeStr).toContain("The old Kingdom of Britania");
    expect(treeStr).toContain("975-09-17");
    expect(treeStr).toContain("The story of King Arthur");
    expect(tree).toMatchSnapshot();
});

it("renders correctly in select mode with disabled radio button", async () => {
    const { findByA11yRole, ...tree } = render(
        <TestLocalizationProvider>
            <ListItem
                author="Merlin"
                country="The old Kingdom of Britania"
                isodate="975-09-17"
                onLongPress={noop}
                onPressInSelectMode={noop}
                selectMode={true}
                selected={false}
                text="The story of King Arthur"
            />
        </TestLocalizationProvider>
    );
    const selectedButton = await findByA11yRole("radio");
    await act(async () => {});

    expect(selectedButton).toBeTruthy();
    // wow, this is incredibly ugly
    const radioButtonFill = (selectedButton.children[0] as any).children[0]
        .children[0].children[0].children[0];
    expect(radioButtonFill).toBe(undefined); // i.e. radio button is off
    expect(tree.toJSON()).toMatchSnapshot();
});

it("renders correctly in select mode with enabled radio button", async () => {
    const { findByA11yRole, debug, ...tree } = render(
        <TestLocalizationProvider>
            <ListItem
                author="Merlin"
                country="The old Kingdom of Britania"
                isodate="975-09-17"
                onLongPress={noop}
                onPressInSelectMode={noop}
                selectMode={true}
                selected={true}
                text="The story of King Arthur"
            />
        </TestLocalizationProvider>
    );
    const selectedButton = await findByA11yRole("radio");

    expect(selectedButton).toBeTruthy();
    // wow, this is incredibly ugly
    const radioButtonFill = (selectedButton.children[0] as any).children[0]
        .children[0].children[0].children[0];
    expect(radioButtonFill).toBeTruthy(); // i.e. radio button is on
    expect(tree.toJSON()).toMatchSnapshot();
});
