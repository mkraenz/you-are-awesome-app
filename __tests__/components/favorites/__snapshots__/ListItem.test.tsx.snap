// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly in normal mode 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "padding": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "marginVertical": 6,
              "paddingLeft": 8,
            },
            Object {
              "flex": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={5}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                      "letterSpacing": 0.5,
                      "lineHeight": 24,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            The story of King Arthur
          </Text>
        </Text>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Merlin
             from 
            The old Kingdom of Britania
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 10,
                    },
                  ],
                ],
              ]
            }
          >
            975-09-17
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly in select mode with disabled radio button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "padding": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "justifyContent": "center",
            "maxWidth": "10%",
          }
        }
      >
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="radio"
          accessibilityState={
            Object {
              "checked": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
              },
            ]
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "borderColor": "rgba(0, 0, 0, 0.54)",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 20,
                "margin": 8,
                "width": 20,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginVertical": 6,
              "paddingLeft": 8,
            },
            Object {
              "flex": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={5}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                      "letterSpacing": 0.5,
                      "lineHeight": 24,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            The story of King Arthur
          </Text>
        </Text>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Merlin
             from 
            The old Kingdom of Britania
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 10,
                    },
                  ],
                ],
              ]
            }
          >
            975-09-17
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly in select mode with enabled radio button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "padding": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "justifyContent": "center",
            "maxWidth": "10%",
          }
        }
      >
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="radio"
          accessibilityState={
            Object {
              "checked": true,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
              },
            ]
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "borderColor": "#03dac4",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 20,
                "margin": 8,
                "width": 20,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "backgroundColor": "#03dac4",
                    "borderRadius": 5,
                    "height": 10,
                    "transform": Array [
                      Object {
                        "scale": 1,
                      },
                    ],
                    "width": 10,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "marginVertical": 6,
              "paddingLeft": 8,
            },
            Object {
              "flex": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={5}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                      "letterSpacing": 0.5,
                      "lineHeight": 24,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            The story of King Arthur
          </Text>
        </Text>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Merlin
             from 
            The old Kingdom of Britania
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 10,
                    },
                  ],
                ],
              ]
            }
          >
            975-09-17
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
